### ISSUE TITLE
Why the "Mov EAX constant" is lifted as "RAX = COPY constant:8" instead of "EAX = COPY constant:4, RAX = INT_ZEXT EAX"

### ISSUE THREAD

**Issue by Muqi-Zou** [NationalSecurityAgency/ghidra#7291]
**Describe the bug**
I notice that ghidra lifts any "Mov EAX x" as "RAX = COPY x", when x is constant (whereas, when x is not a constant, it lifts "Mov EAX x" as "EAX = COPY/Load X, RAX = INT_ZEX EAX"). I am not saying this constant lifting is semantically incorrect, since "[Most (if not all) 32-bit MOVs and ALU operations clear bits 32 through 63 of the destination register operand. ](https://stackoverflow.com/questions/11822646/can-mov-eax-0x1-always-be-used-instead-of-mov-rax-0x1)".  However, this direct copy to 64-bit register may cause the type of constant cannot be accurately identified during the decompilation since the INT_ZEXT is one of the key operations trigger the subvarflow analysis (i.e., RuleSubZext). As you can see from the screenshot since the constant is lifted as 8 bytes, the return variable does not have enough information to be recovered as an int.

Hence, my question is what is the reason for ghidra to discard this size information during the lifting?

**Screenshots**
![Screenshot from 2024-12-17 03-01-37](https://github.com/user-attachments/assets/92a5523e-bcdf-417d-9da0-f79d41d5cd1d)

**Attachments**
[chcon_O0.zip](https://github.com/user-attachments/files/18162196/chcon_O0.zip)

**Environment (please complete the following information):**
Ghidra 11.3

**Comment by GhidorahRex** [comment]
Absent any additional information, there's no way of knowing if the original meaning was a long or an int. `mov EAX 0xffffffff` is semantically equivalent to `mov RAX 0xffffffff`. Since they're equivalent, was the original intent to store a 32-bit value or a 64-bit value?

Just to check, I updated the pcode for `mov r32, imm32` to use the 32-bit register and zext to the 64-bit register and the analysis of your function here is exactly the same. I would be interested in seeing any examples where there was a clear difference in the analysis, but I haven't seen one yet.

You can validate it yourself with changing this line:
```
:MOV CRmr32,imm32   is vexMode=0 & opsize=1 & row=11 & page=1 & CRmr32; imm32       { CRmr32 = imm32; }
```
to this:
```
:MOV Rmr32,imm32   is vexMode=0 & opsize=1 & row=11 & page=1 & Rmr32 & Rmr64; imm32       { Rmr32 = imm32; Rmr64 = zext(Rmr32); }
```

**Comment by Muqi-Zou** [comment]
> Absent any additional information, there's no way of knowing if the original meaning was a long or an int.

I do not agree with this statement. When seeing `mov eax constantA`, I personally believe this is already a strong hint that this constantA's size should not be larger than 4 bytes. If the constantA is larger like 8 bytes, assembler should throw error. That is, for instance, we should never see `mov eax, 0x1234567890ABCDEF` or if this invalid statement is generated, the 32 through 63 bits should be cleared to 0 (feel free to correct me if this is wrong). Hence, converting `mov eax, constant` to `mov rax, constant` at the lifting stage loses the hint for the constant's size.

For the `RuleSubZext`, adding int_zext in p-code will not cause the type directly change, since inside `RuleSubZext`, it only conducts the subvarflow analysis when the extended varnode is defined within `CPUI_SUBPIECE` operation or `CPUI_INT_RIGHT`(check [here](https://github.com/NationalSecurityAgency/ghidra/blob/8dd0ea698ac8b9bc6a8d973d882c71037714e81e/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc#L4964)). In the other world, we may need to rewrite the this rule to make the type be correctly identified.

Btw, I cannot compile after above provided change, but I can compile with following change, not sure it is because I am using older version of ghidra.
```
:MOV Rmr32,imm32   is vexMode=0 & opsize=1 & row=11 & page=1 & Rmr32 &  check_Rmr32_dest; imm32       { Rmr32 = imm32; build  check_Rmr32_dest; }
```

**Comment by GhidorahRex** [comment]
> > Absent any additional information, there's no way of knowing if the original meaning was a long or an int.
> 
> I do not agree with this statement. When seeing `mov eax constantA`, I personally believe this is already a strong hint that this constantA's size should not be larger than 4 bytes. If the constantA is larger like 8 bytes, assembler should throw error. That is, for instance, we should never see `mov eax, 0x1234567890ABCDEF` or if this invalid statement is generated, the 32 through 63 bits should be cleared to 0 (feel free to correct me if this is wrong). Hence, converting `mov eax, constant` to `mov rax, constant` at the lifting stage loses the hint for the constant's size.
> 
> For the `RuleSubZext`, adding int_zext in p-code will not cause the type directly change, since inside `RuleSubZext`, it only conducts the subvarflow analysis when the extended varnode is defined within `CPUI_SUBPIECE` operation or `CPUI_INT_RIGHT`(check [here](https://github.com/NationalSecurityAgency/ghidra/blob/8dd0ea698ac8b9bc6a8d973d882c71037714e81e/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc#L4964)). In the other world, we may need to rewrite the this rule to make the type be correctly identified.
> 
> Btw, I cannot compile after above provided change, but I can compile with following change, not sure it is because I am using older version of ghidra.
> 
> ```
> :MOV Rmr32,imm32   is vexMode=0 & opsize=1 & row=11 & page=1 & Rmr32 &  check_Rmr32_dest; imm32       { Rmr32 = imm32; build  check_Rmr32_dest; }
> ```

I don't have any comment regarding the change to the rules pieces, but I think you misunderstood my comment about the instruction itself.

`mov eax, 0x0` for instance is semantically equivalent to `mov rax, 0x0` - there's NO difference between them. Compilers often use the 32-bit version to save bytes. The same is true for any constants which fit in a 32-bit variable. So although `mov rax, 0xffffffffffffffff` is definitely a 64-bit load, we don't know outside of any additional information if `0xffffffff` is supposed to be -1 in 32-bits or 0x00000000ffffffff in 64-bits. Compilers will prefer `mov eax, 0xffffffff` over `mov rax, 0x00000000ffffffff`

**Comment by Wall-AF** [comment]
> `mov eax, 0x0` for instance is semantically equivalent to `mov rax, 0x0` - there's NO difference between them. Compilers often use the 32-bit version to save bytes. The same is true for any constants which fit in a 32-bit variable. So although `mov rax, 0xffffffffffffffff` is definitely a 64-bit load, we don't know outside of any additional information if `0xffffffff` is supposed to be -1 in 32-bits or 0x00000000ffffffff in 64-bits. Compilers will prefer `mov eax, 0xffffffff` over `mov rax, 0x00000000ffffffff`

In that case maybe a "user option" could be enabled to facilitate such cases?

**Comment by Muqi-Zou** [comment]
After some research about compiler/assembler, I agree with the statement that we cannot simply identify the number of bytes once we see `mov eax`. The reason behind it is very complicated, here are some good explanations (about [nasm/yasm](https://stackoverflow.com/a/48597025),  about [gcc](https://stackoverflow.com/a/38314331),  about [cpu](https://stackoverflow.com/a/41574531), and about [security](https://stackoverflow.com/a/40535549)). In short, because of the above reasons, the `mov eax` we see by the disassembler may come from `mov rax`, `mov rax`,  `mov ax` or even  `mov al`. However, IMHO, these analysis/guesses about constant size should be conducted in the decompiler instead of the lifter, since my experience is that p-code is an explicit IR language. If we propose that the p-code lifter should be explicit, `mov eax` needs to be lifted to `eax = copy; rax = int_zex eax`. However, this is up to the decompiler developers' opinions. Hence, feel free to close this issue.
